modules/graceful/manager_unix.go

```diff
@@ -139,6 +139,7 @@ func (g *Manager) handleSignals(ctx context.Context) {
 				if err := log.ReleaseReopen(); err != nil {
 					log.Error("Error whilst releasing and reopening logs: %v", err)
 				}
+				// TODO audit.ReleaseReopen
```

Comment:
@zeripath I can't call this method here because of import cycles. What do you think about this addition?

```diff
--- a/modules/graceful/manager_unix.go
+++ b/modules/graceful/manager_unix.go
@@ -41,6 +41,7 @@ type Manager struct {
        toRunAtShutdown  []func()
        toRunAtHammer    []func()
        toRunAtTerminate []func()
+       toRunAtSignal    []func(os.Signal)
 }
 
 func newGracefulManager(ctx context.Context) *Manager {
@@ -110,6 +111,21 @@ func (g *Manager) start(ctx context.Context) {
        }
 }
 
+// RunAtSignal creates a go-routine to run the provided function when receiving a signal
+func (g *Manager) RunAtSignal(fn func(os.Signal)) {
+       g.lock.Lock()
+       defer g.lock.Unlock()
+       g.toRunAtSignal = append(g.toRunAtSignal,
+               func(signal os.Signal) {
+                       defer func() {
+                               if err := recover(); err != nil {
+                                       log.Critical("PANIC during RunAtSignal: %v\nStacktrace: %s", err, log.Stack(2))
+                               }
+                       }()
+                       fn(signal)
+               })
+}
+
 func (g *Manager) handleSignals(ctx context.Context) {
        ctx, _, finished := process.GetManager().AddTypedContext(ctx, "Graceful: HandleSignals", process.SystemProcessType, true)
        defer finished()
@@ -154,6 +170,9 @@ func (g *Manager) handleSignals(ctx context.Context) {
                        default:
                                log.Info("PID %d. Received %v.", pid, sig)
                        }
+                       for _, fn := range g.toRunAtSignal {
+                               go fn(sig)
+                       }
                case <-ctx.Done():
                        log.Warn("PID: %d. Background context for manager closed - %v - Shutting down...", pid, ctx.Err())
                        g.DoGracefulShutdown()
```
This would allow to register the callbacks without dependency problems. The downside is, this is only used for unix system so the callback-registration needs to happen from platform specific files.

Reply:
I think the problem is:

it shouldn't be like this:

```
func Audit.Init() {
    graceful.RunAtShudown(...)
}
```

But it should make the global initializer do init&shutdown:

```
func Global.Init() {
    audit.Init()
    graceful.RunAtShudown(audit.Shuwdown)
}
```

Reply:
Is there followup here or is it acceptable for now?

Reply:
With #24726 there is no need for the ToDo anymore.